common/scala/src/main/scala/whisk/core/database/memory/MemoryAttachmentStore.scala

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.database.memory
+
+import akka.actor.ActorSystem
+import akka.event.Logging.ErrorLevel
+import akka.http.scaladsl.model.ContentType
+import akka.stream.{ActorMaterializer, OverflowStrategy, QueueOfferResult}
+import akka.stream.scaladsl.{Keep, Sink, Source}
+import akka.util.{ByteString, ByteStringBuilder}
+import whisk.common.{Logging, StartMarker, TransactionId}
+import whisk.common.LoggingMarkers.DATABASE_ATT_GET
+import whisk.common.LoggingMarkers.DATABASE_ATT_SAVE
+import whisk.common.LoggingMarkers.DATABASE_ATT_DELETE
+import whisk.core.database._
+import whisk.core.entity.{DocId, DocInfo}
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.reflect.ClassTag
+
+object MemoryAttachmentStoreProvider extends AttachmentStoreProvider {
+  override def makeStore[D <: DocumentSerializer: ClassTag]()(implicit actorSystem: ActorSystem,
+                                                              logging: Logging,
+                                                              materializer: ActorMaterializer): AttachmentStore =
+    new MemoryAttachmentStore(implicitly[ClassTag[D]].runtimeClass.getSimpleName.toLowerCase)
+}
+
+class MemoryAttachmentStore(dbName: String)(implicit system: ActorSystem,
+                                            logging: Logging,
+                                            materializer: ActorMaterializer)
+    extends AttachmentStore {
+
+  override protected[core] implicit val executionContext: ExecutionContext = system.dispatcher
+
+  private sealed trait Action {
+    def doc: DocInfo
+  }
+  private case class AddAction(doc: DocInfo, name: String, attachment: Attachment) extends Action
+  private case class DelAction(doc: DocInfo) extends Action
+  private case class Attachment(bytes: ByteString, contentType: ContentType)
+
+  @volatile
+  private var attachments: Map[String, Attachment] = Map()
```

Comment:
in particular, `attachments` here appears to be intended as a cache (would be nice to mention in the class docs).
I have two concerns with this Map - while you use volatile, I think you might want to consider using a TrieMap instead http://www.scala-lang.org/api/2.12.1/scala/collection/concurrent/TrieMap.html.

Second, attachments can be large. This Map has no bound on its size and can grow to OOM the controller.
Instead, maybe it's possible to tie into the cache implementation that already exists which uses soft references, and furthermore has a time to live bound and max key count (still not ideal for attachments where a size bound is preferred).

Reply:
Did not knew about `TrieMap`. Thanks for the tip!. Now the `MemoryAttachmentStore` implementation is much simpler. 

> Second, attachments can be large. This Map has no bound on its size and can grow to OOM the controller

Purpose of memory store is to facilitate testing and there also each test would have its own instance. So we can keep it simple with assumption that tests would not be storing very large binaries in it.

To make it more explicit (i.e. it is meant for test) I have moved the logic to tests module